src/components/checkbox/checkbox.scss

```diff
@@ -0,0 +1,460 @@
+ */
+@keyframes md-checkbox-checked-indeterminate-mixedmark {
+  from {
+    animation-timing-function: $md-linear-out-slow-in-timing-function;
+    opacity: 0;
+    transform: rotate(-45deg);
+  }
+
+  to {
+    opacity: 1;
+    transform: rotate(0deg);
+  }
+}
+
+/**
+ * Rotates and fades out the mixedmark when the checkbox goes from indeterminate -> checked. This
+ * animation, similar to md-checkbox-indeterminate-checked-checkmark, helps provide an illusion
+ * of "morphing" from mixedmark -> checkmark.
+ */
+@keyframes md-checkbox-indeterminate-checked-mixedmark {
+  from {
+    animation-timing-function: $_md-checkbox-indeterminate-checked-easing-function;
+    opacity: 1;
+    transform: rotate(0deg);
+  }
+
+  to {
+    opacity: 0;
+    transform: rotate(315deg);
+  }
+}
+
+/**
+ * Horizontally collapses and fades out the mixedmark when the checkbox goes from indeterminate ->
+ * unchecked.
+ */
+@keyframes md-checkbox-indeterminate-unchecked-mixedmark {
+  0% {
+    animation-timing-function: linear;
+    opacity: 1;
+    transform: scaleX(1);
+  }
+
+  32.8%, 100% {
+    opacity: 0;
+    transform: scaleX(0);
+  }
+}
+
+/**
+ * Applied to elements that cover the checkbox's entire inner container.
+ */
+%md-checkbox-cover-element {
+  bottom: 0;
+  left: 0;
+  position: absolute;
+  right: 0;
+  top: 0;
+}
+
+/**
+ * Applied to elements that are considered "marks" within the checkbox, e.g. the checkmark and
+ * the mixedmark.
+ */
+%md-checkbox-mark {
+  $width-padding-inset: 2 * $md-checkbox-border-width;
+
+  width: calc(100% - #{$width-padding-inset});
+  will-change: opacity, transform;
+}
+
+/**
+ * Applied to elements that appear to make up the outer box of the checkmark, such as the frame
+ * that contains the border and the actual background element that contains the marks.
+ */
+%md-checkbox-background {
+  @extend %md-checkbox-cover-element;
+  border-radius: 2px;
+  box-sizing: border-box;
+  pointer-events: none;
+}
+
+.md-checkbox {
+  cursor: pointer;
+
+  &:focus {
+    // TODO(traviskaufman): Add ink ripple on focus state, once ripple is implemented.
+    outline: none;
+  }
+}
+
+.md-checkbox-layout {
+  align-items: baseline;
+  display: inline-flex;
+}
+
+.md-checkbox-inner-container {
+  display: inline-block;
+  height: $md-checkbox-size;
+  line-height: 0;
+  margin: auto;
+  margin-right: $md-checkbox-item-spacing;
+  order: 0;
+  position: relative;
+  vertical-align: middle;
+  white-space: nowrap;
+  width: $md-checkbox-size;
+
+  [dir="rtl"] & {
+    margin:{
+      left: $md-checkbox-item-spacing;
+      right: auto;
+    }
+  }
+}
+
+.md-checkbox-frame {
+  @extend %md-checkbox-background;
+
+  background-color: transparent;
+  border: $md-checkbox-border-width solid $md-checkbox-border-color;
+  transition: border-color $md-checkbox-transition-duration $md-linear-out-slow-in-timing-function;
+  will-change: border-color;
+}
+
+.md-checkbox-background {
+  @extend %md-checkbox-background;
```

Comment:
Nitpick: do you think it's confusing that the class and the placeholder share the same name?  Since the comment above cites the "outer box", that might be a more descriptive name for the placeholder.